### PR TITLE
[Diagnostics] Provide contextual type when diagnosing invalid if-exp

### PR DESCRIPTION
**Pull request by mkchoi212**
This pull request now provides better diagnostics (with fixits) when invalid type/member is provided as a return value of a ternary operator.

Resolves [SR-910](https://bugs.swift.org/browse/SR-910).

cc @xedin

### CODE REVIEW

File: lib/Sema/CSDiag.cpp [hunk 1]
```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
 }
 
 bool FailureDiagnosis::visitIfExpr(IfExpr *IE) {
-  auto typeCheckClauseExpr = [&](Expr *clause) -> Expr * {
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertTypePurpose) -> Expr * {
```
Comment (CodaFi) [inline lib/Sema/CSDiag.cpp]: You should ask the constraint system for the purpose of the contextual type 

`cs.getContextualTypePurpose()`

File: test/Constraints/diagnostics.swift [hunk 2]
```diff
@@ -188,6 +188,23 @@ func r17224804(_ monthNumber : Int) {
   let monthString = (monthNumber <= 9) ? ("0" + monthNumber) : String(monthNumber)
 }
 
+// <rdar://problem/33768959> QoI: invalid diagnostics when input to ternary operator does not exist
+func r33768959() {
+	enum E {
```
Comment (CodaFi) [inline test/Constraints/diagnostics.swift]: Indentation here is tabs, please use spaces to match the rest of this file

File: test/Constraints/diagnostics.swift [hunk 3]
```diff
@@ -188,6 +188,23 @@ func r17224804(_ monthNumber : Int) {
...
+        case LessThan
+        case EqualTo
+        // expected-note @+1 {{did you mean 'GreaterThan'?}}
+        case GreaterThan
+	}
+
+	func compare(_ lhs: Int, _ rhs: Int) -> E {
+        guard lhs != rhs else { return .EqualTo }
```
Comment (CodaFi) [inline test/Constraints/diagnostics.swift]: Indent every syntactic level with two spaces.

Comment (CodaFi) [inline test/Constraints/diagnostics.swift]: Also, this part isn't really relevant to diagnostic emission and could be removed

File: lib/Sema/CSDiag.cpp (hunk 1 above)
Comment (mkchoi212) [inline lib/Sema/CSDiag.cpp]: `CS.getContextualTypePurpose()` of a `condExpr` evaluates to `CTP_ReturnStmt`, causing the compiler  to emit the old message and exiting
`error: cannot convert return expression of type 'Bool' to return type 'E'`. 

Should I replace `condExpr`'s CTP explicitly to `CTP_Unused`?

File: test/Constraints/diagnostics.swift (hunk 3 above)
Comment (xedin) [inline test/Constraints/diagnostics.swift]: @mkchoi212 Just to clarify a little bit, we are trying to pick the snippets of code we use in the test suite as small as possible so it's worth to reduce the example from issue even further and delete all irrelevant parts.

File: lib/Sema/CSDiag.cpp (hunk 1 above)
Comment (xedin) [inline lib/Sema/CSDiag.cpp]: @mkchoi212 @CodaFi is right, if we want to provide contextual type to `typeCheckClauseExpr` we need to use both type and its purpose from the original constraint system. It's just so happens that in the snippet attached to the patch it is CTP_ResultStmt, but it might easily be e.g. CTP_AssignSource. 

To properly fix the problem at hand (and you should also add additional test-case e.g. `let _: E = lhs < rhs ? .LessThan : .GreaterTham`) you'll have to make changes to `diagnoseMemberFailures` method which should be called when `.GreaterTham` is type-checked separately, since it doesn't matter what is the purpose of the contextual type when we use to to resolve `UnresolvedMemberExpr` it would work all different kinds of expressions.

Comment (CodaFi) [inline lib/Sema/CSDiag.cpp]: You should type check the condition as unused then and just use the purpose for the branches.

Comment (mkchoi212) [inline lib/Sema/CSDiag.cpp]: @xedin The code is not breaking at any of `diagnoseMemberFailures` within `CSDiag.cpp`. Instead, should I try to add more `tryRawRepresentableFixIts` to a switch statement within `diagnoseContextualConversionError`?

Comment (mkchoi212) [inline lib/Sema/CSDiag.cpp]: @xedin Nothing is breaking under `diagnoseMemberFailures` :( 
Should we try to add more `try____FixIts` under a switch statement in `diagnoseContextualConversionError` for `CTP_ReturnStmt`?

Comment (xedin) [inline lib/Sema/CSDiag.cpp]: That is exactly the problem - error is not detected at the right place and it is not contextual, it's related to the member lookup, which should be diagnosed by `visitUnresolvedMemberExpr` and more specifically `diagnoseMemberFailures`. You need to try and figure out how to make that happen to properly fix diagnostic.

File: test/Constraints/diagnostics.swift [hunk 4]
```diff
@@ -188,6 +188,23 @@ func r17224804(_ monthNumber : Int) {
...
+        // expected-error @+1 {{type 'E' has no member 'GreaterTham'}}
+        return lhs < rhs ? .LessThan : .GreaterTham
```
Comment (CodaFi) [inline test/Constraints/diagnostics.swift]: Yes.  This test specifically requires this typo so lookup fails.

**Comment by slavapestov** [comment]
A few more test cases to look at:
- https://bugs.swift.org/browse/SR-2326
- https://bugs.swift.org/browse/SR-3839 
- https://bugs.swift.org/browse/SR-2336

**Comment by mkchoi212** [comment]
@slavapestov sorry for not having updated anything for awhile. I've been busy with stuff! Do you have the fix or should I keep going at it?

**Comment by CodaFi** [comment]
Yes.  You should use unused for the Boolean condition and ask the constraint system for the purpose of the contextual type at the branches.

**Comment by mkchoi212** [comment]
@CodaFi Ok great! I guess I just overcomplicating things. I will check if there are any pre-existing tests for this, make some changes and commit them later today.

File: lib/Sema/CSDiag.cpp [hunk 5]
```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
...
+      // Provide proper contextual type when type conversion is specified.
+      auto type = convertTypePurpose == CTP_Unused ? Type() : CS.getContextualType();
+      return typeCheckChildIndependently(clause, type, convertTypePurpose, TCCOptions(),
```
Comment (xwu) [inline lib/Sema/CSDiag.cpp]: Drive-by nit: please adhere to 80-column lines and indent two spaces in this file.

**Comment by CodaFi** [comment]
Cool!  Looks great.

@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

File: lib/Sema/CSDiag.cpp [hunk 6]
```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
...
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
-                                       nullptr, false);
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertPurpose) -> Expr * {
+      // Provide proper contextual type when type conversion is specified.
+      auto type = convertPurpose == CTP_Unused ? Type() : CS.getContextualType();
+      return typeCheckChildIndependently(clause, type, convertPurpose, TCCOptions(),
+                                         nullptr, false);
   };
-
+
```
Comment (xedin) [inline lib/Sema/CSDiag.cpp]: Please remove whitespace.

File: lib/Sema/CSDiag.cpp [hunk 7]
```diff
@@ -7200,18 +7200,19 @@ bool FailureDiagnosis::visitBindOptionalExpr(BindOptionalExpr *BOE) {
 }
 
 bool FailureDiagnosis::visitIfExpr(IfExpr *IE) {
-  auto typeCheckClauseExpr = [&](Expr *clause) -> Expr * {
-    return typeCheckChildIndependently(clause, Type(), CTP_Unused, TCCOptions(),
-                                       nullptr, false);
+  auto typeCheckClauseExpr = [&](Expr *clause, ContextualTypePurpose convertPurpose) -> Expr * {
```
Comment (xedin) [inline lib/Sema/CSDiag.cpp]: Please pass both type and purpose as parameters (you can have type default to Type() and purpose to CTP_Unused), I am trying to reduce usage of CS.getContextualType in places like this, it could be computed right before the call.

**Review by xedin — CHANGES_REQUESTED** [review]
Overall LGTM, only requires some minor changes now.

**Review by xedin — CHANGES_REQUESTED** [review]
Ok, one last thing - please squash everything into a single commit and mention `Resolves: SR-xxxx` in the commit message, also please use `clang-format` on changes to make sure that everything is formatted properly.

**Comment by mkchoi212** [comment]
@xedin Is there a script to run the `clang-format` and `swift-format`? `../build/Ninja-DebugAssert/llvm-macosx-x86_64/bin/clang-format -i -style=LLVM lib/Sema/CSDiag.cpp ` is making lots of changes.`swift-format` is also showing parse errors.

**Comment by xedin** [comment]
There is `git-clang-format` script in clang/utils you can use right before commit to format only staged changes. I just noticed that commit should also say [Diagnostics] instead of [Sema] because changes are related to diagnostics.

**Comment by mkchoi212** [comment]
@xedin Changes have been made. Please review :)

**Comment by xedin** [comment]
@mkchoi212 Perfect, thank you!

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Comment by xedin** [comment]
@swift-ci please smoke test and merge

**Comment by mkchoi212** [comment]
@CodaFi @slavapestov @xedin Thank you guys for helping me complete my first PR to Swift! I learned a lot during the process. Moving on to the next bug!
